component/loki/write/write.go

```diff
@@ -119,11 +119,14 @@ func (c *Component) Run(ctx context.Context) error {
 		case <-ctx.Done():
 			return nil
 		case entry := <-c.receiver.Chan():
+			c.mut.RLock()
 			select {
 			case <-ctx.Done():
+				c.mut.RUnlock()
 				return nil
 			case c.sink.Chan() <- entry:
```

Comment:
Something about this feels like it could be made a bit cleaner. Thoughts on having an unchanging channel thats instantiated on new func that feeds the sink? The the lock can be moved out of here. Note this might be uglier/worse than what we have here.

Reply:
You were thinking something like:
```

func (c *Component) runFeeder(ctx context.Context) {
	for entry := range c.feedingChannel {
		c.mut.RLock()
		select {
		case <-ctx.Done():
			// exit promptly
			c.mut.RUnlock()
			return
		case c.sink.Chan() <- entry:
		}
		c.mut.RUnlock()
	}
}

// Run implements component.Component.
func (c *Component) Run(ctx context.Context) error {
	// defer corresponding cleanups
	defer close(c.feedingChannel)

	for {
		select {
		case <-ctx.Done():
			return nil
		case entry := <-c.receiver.Chan():
			select {
			case <-ctx.Done():
				return nil
			case c.feedingChannel <- entry:
			}
		}
	}
}
```

IMO the current impl looks cleaner, but adding that extra channel might cost no perf impact I guess. However, it means adding even more one go-routine to run that feeding channel.